### PR TITLE
Update CassandraRowStream with Vert.x 3.6.0 ReadStream changes

### LINKED ISSUE #8: Update CassandraRowStreamImpl with implementing fetch method
According to https://github.com/vert-x3/wiki/wiki/Vert.x-Roadmap#stream-improvements, `CassandraRowStreamImpl` should have `fetch` method

### PR DESCRIPTION
**Pull request by Sammers21**
Fixes #8

### CODE REVIEW

**Comment by Sammers21** [comment]
Do we really need to use it? I guess the code would be more simple, but there are will be an overhead: we will keep an intermediate queue.

Now it is implemented without any intermediate storage.

**Comment by vietj** [comment]
ah ok, how can you implement it without an intermediate queue ?

**Comment by vietj** [comment]
I've looked at the impl, I think we should use the queue to simplify code, the queue will be fine, wether items are in the actual cassandra result set of in the uque, does not make much a difference.

**Comment by Sammers21** [comment]
@vietj , I have just reimplemented it with the Vert.x queue. Can you review?

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java [hunk 1]
```diff
@@ -47,33 +51,34 @@ public CassandraRowStreamImpl(ResultSet result, Vertx vertx) {
...
     return this;
   }
 
   @Override
-  public synchronized CassandraRowStream resume() {
-    paused = false;
-    fireStream();
+  public CassandraRowStream resume() {
+    internalQueue.resume();
+    fire();
```
Comment (vietj) [inline src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: fire() should actually not be called

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java [hunk 2]
```diff
@@ -84,47 +89,60 @@ public synchronized CassandraRowStream endHandler(Handler<Void> handler) {
...
-        } else {
-          context.runOnContext(v -> {
-            synchronized (thisStream) {
-              if (!paused) {
-                rowHandler.handle(resultSetIterator.next());
-              }
-            }
-          });
+  @Override
+  public synchronized CassandraRowStream fetch(long l) {
```
Comment (vietj) [inline src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: I think we should just delegate to the queue, it's not an issue if that does not match the number of items. the number of items is actually what is read when we need to read new rows, and it's independant of the fetch value

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Sure

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Ah, I got the idea. It seems that previously, I misunderstood what the method should actually do.

**Comment by Sammers21** [comment]
@vietj, can you review one more time, please?

**Comment by Sammers21** [comment]
@vietj , ping

**Comment by Sammers21** [comment]
@tsegismont , can you review this PR?

**Comment by Sammers21** [comment]
@tsegismont , ping

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java [hunk 3]
```diff
@@ -84,47 +88,55 @@ public synchronized CassandraRowStream endHandler(Handler<Void> handler) {
...
+    }
+
+    if (internalQueue.isWritable()) {
+      fetchAndCallOneMoreTime();
+    }
+  }
+
+  private void fetchAndCallOneMoreTime() {
+    fetch().setHandler(v -> {
+      if (v.succeeded()) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: When #4 will be merged, this should be wrapped in `executeFromIO`

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java [hunk 4]
```diff
@@ -84,47 +88,55 @@ public synchronized CassandraRowStream endHandler(Handler<Void> handler) {
...
+  @Override
+  public synchronized CassandraRowStream fetch(long l) {
+    internalQueue.take(l);
+    return this;
+  }
+
+  private synchronized void fire() {
+    int availableWithoutFetching = datastaxResultSet.getAvailableWithoutFetching();
+    boolean isFetched = datastaxResultSet.isFullyFetched();
+    if (availableWithoutFetching == 0 && isFetched) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: I believe it would an if/else block would be easier to read than the double if.

Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Not sure if we can replace these 2 if blocks with a single if/else block, but I tried to make to more readable. Is it fine?

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java [hunk 5]
```diff
@@ -31,9 +32,12 @@
  */
 public class CassandraRowStreamImpl implements CassandraRowStream {
 
+  private static final Long STOP_STREAM_FETCH_NUMBER = 0L;
```
[the remark this answers is not in the document]
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Done

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Sure

**Comment by Sammers21** [comment]
@tsegismont, can you review one more time, please?

File: src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraRowStreamImpl.java]: Let's mergre this PR first and then i wil update the #4 PR.

**Comment by tsegismont** [comment]
@Sammers21 thank you!

**Pull request merged**